scipy/optimize/optimize.py

```diff
@@ -2397,20 +2397,81 @@ def bracket(func, xa=0.0, xb=1.0, args=(), grow_limit=110.0, maxiter=1000):
     return xa, xb, xc, fa, fb, fc, funcalls
 
 
-def _linesearch_powell(func, p, xi, tol=1e-3):
+def _m_min(a, b):
+    """ Return min(a, b) """
+    if a is None:
+        return b
+    elif b is None:
+        return a
+    else:
+        return min(a, b)
+
+
+def _m_max(a, b):
+    """ Return max(a, b) """
+    if a is None:
+        return b
+    elif b is None:
+        return a
+    else:
+        return max(a, b)
+
+
+def _line_for_search(x0, alpha, lower_bound, upper_bound):
+    """
+    x0 is the vector representing the current location
+    alpha is the unit vector representing the direction
+    
+    ie the direction is along the line from x0 to alpha
+
+    lower_bound is a list/array of the lower bounds for each parameter in x0.
+    upper_bound is a list/array of the upper bounds for each parameter in x0.
+    
+    returns (lmin, lmax), the bounds for 
+            lower_bound[i] <= x0_i+alpha_i*l <= upper_bound[i] 
+        for all i.
+    """
+    # figure out how far forward we can go before we are out of bounds
+    # for one of the params
+    lmin, lmax = None, None
+    for i in range(len(alpha)):
```

Comment:
I suspect this can be vectorized.

Reply:
Done. Let me know if you think it is good.

Reply:
Suggestion for simplification below.

Reply:
Thanks, great idea!